### PR TITLE
Enable Portable ThreadPool events to be written using NativeRuntimeEventSource

### LINKED ISSUE #45384: Support Managed Threadpool Diagnostics
The goal is to maintain parity with existing diagnostics for the native thread pool.  Currently the concrete work for this item is:

- [ ] Make NativeRuntimeEventSource.cs as a general way of sending runtime events from managed side, as opposed to what it is now
- [x] Make sure the events being fired from the managed ThreadPool is being fired correctly.
- [ ] Fix EventListener deadlock that can occur as part of enabling managed threadpool events through EventListener 
- [ ] Make sure EventListeners can subscribe to managed threadpool events
- [ ] Test LTTng events can still get fired from managed threadpool under the Microsoft-Windows-DotNETRuntime provider

### PR DESCRIPTION
**Pull request by sywhang**
This removes the temporary solution that the portable ThreadPool was using to emit events from managed side. Specifically, the following changes were made: 
* Change the GUID of NativeRuntimeEventSource to match the native provider (Microsoft-Windows-DotNETRuntime)
* Remove the hacks in EventPipe and EventSource that existed for the temporary solution
* Add endpoints in XplatEventSourceLogger that correspond to each managed event that acts as a sink for forwarding events to LTTng
*  Change the genRuntimeEventSources.py script to not generate events for the hand-written ThreadPool events. 

Fix #45384

cc @brianrob @josalem @kouvel @noahfalk @tommcdon

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs [hunk 1]
```diff
@@ -3801,16 +3801,6 @@ public class EventListener : IDisposable
         /// </summary>
         public event EventHandler<EventWrittenEventArgs>? EventWritten;
 
-        static EventListener()
-        {
-#if FEATURE_PERFTRACING
```
Comment (brianrob) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs:3806]: Is `NativeRuntimeEventSource` now guaranteed to always be initialized, even just for a console app?

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/TraceLogging/XplatEventLogger.cs [hunk 2]
```diff
@@ -160,12 +271,6 @@ private static void AppendByteArrayAsHexString(StringBuilder builder, byte[] byt
 
         protected internal override void OnEventSourceCreated(EventSource eventSource)
         {
-            // Don't enable forwarding of NativeRuntimeEventSource events.`
-            if (eventSource.GetType() == typeof(NativeRuntimeEventSource))
```
Comment (brianrob) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/TraceLogging/XplatEventLogger.cs:164]: If I recall correctly, this is here because otherwise events that are written by the runtime end up getting pushed to this `EventListener` and then back down to LTTng.  This only happens when COMPlus_EnableEventLog=1, and then you only see it if you're capturing via LTTng I believe.  The details are a bit fuzzy...

File: src/libraries/System.Private.CoreLib/src/System/Threading/NativeRuntimeEventSource.PortableThreadPool.cs [hunk 3]
```diff
@@ -134,6 +119,9 @@ private unsafe void WriteThreadEvent(int eventId, uint numExistingThreads)
             if (IsEnabled(EventLevel.Informational, Keywords.ThreadingKeyword))
```
Comment (brianrob) [inline src/libraries/System.Private.CoreLib/src/System/Threading/NativeRuntimeEventSource.PortableThreadPool.cs:119]: I may have missed it, but what ensures that `IsEnabled` returns true?

**Review by brianrob — COMMENTED** [review]
Awesome to see this.  A few things that come to mind.

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/TraceLogging/XplatEventLogger.cs [hunk 4]
```diff
@@ -44,6 +44,117 @@ internal class XplatEventLogger : EventListener
         [DllImport(RuntimeHelpers.QCall, CharSet = CharSet.Unicode)]
         private static extern void LogEventSource(int eventID, string? eventName, string eventSourceName, string payload);
 
+        [DllImport(RuntimeHelpers.QCall)]
+        private static extern void LogThreadPoolWorkerThreadStart(uint ActiveWorkerThreadCount, uint RetiredWorkerThreadCount, ushort ClrInstanceID);
+
+        public static void ThreadPoolWorkerThreadStart(uint ActiveWorkerThreadCount, uint RetiredWorkerThreadCount, ushort ClrInstanceID)
+        {
+            LogThreadPoolWorkerThreadStart(ActiveWorkerThreadCount, RetiredWorkerThreadCount, ClrInstanceID);
+        }
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/TraceLogging/XplatEventLogger.cs:53]: This level of indirection is required?

File: src/libraries/System.Private.CoreLib/src/System/Threading/NativeRuntimeEventSource.PortableThreadPool.cs [hunk 5]
```diff
@@ -183,6 +171,10 @@ private unsafe void WriteThreadEvent(int eventId, uint numExistingThreads)
             data[3].Size = sizeof(ushort);
             data[3].Reserved = 0;
             WriteEventCore(55, 4, data);
+
+#if FEATURE_EVENTSOURCE_XPLAT
+            XplatEventLogger.ThreadPoolWorkerThreadAdjustmentAdjustment(AverageThroughput, NewWorkerThreadCount, Reason, ClrInstanceID);
+#endif
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Threading/NativeRuntimeEventSource.PortableThreadPool.cs:177]: I could use some education... why is this needed, rather than for example registering a listener that routes events to LTTng?

Also, moving forward is LTTng a path we're still excited about / supporting?  I assumed with event pipe and the rest of the dotnet tooling around tracing cross-platform that we'd be deprecating this support.

Comment (brianrob) [reply src/libraries/System.Private.CoreLib/src/System/Threading/NativeRuntimeEventSource.PortableThreadPool.cs:177]: We do still want to support LTTng, as we're seeing some amount of uptake across other Microsoft products, and we want to be able to participate in this.  That said, this code is essentially making it look like the thread pool is still implemented in native code.  So rather than using an EventSource in managed code, this is invoking the native FireETW* functions, which result in calls to EventPipe, ETW, and/or LTTng.  So, this code is actually platform agnostic, and takes advantage of the platform-specific logger code in the runtime itself.

That said, thinking about this a bit further, I am wondering if there's something that we can't do here in a more generic direction.  Rather than having to create a new method and QCALL for every event, can we just use the `NativeRuntimeEventSource` as is, and modify `EventPipe` to be OK with multiple provider registrations (at least for the runtime provider)?  I could believe that over time there will be more components that move to managed code, and we'll want to keep logging the same events.  ETW will already allow multiple provider registrations.  LTTng might need something special - possibly some pre-generated code that would translate between the `EventSource` and the LTTng tracepoints.  But, I'd be tempted to say that we shouldn't do something special for it, and just treat the events like any other `EventSource` events with respect to LTTng.

File: src/coreclr/scripts/genRuntimeEventSources.py [hunk 6]
```diff
@@ -147,6 +155,13 @@ def generateEvent(eventNode, providerNode, outputFile, stringTable):
     outputFile.write(")\n")
     writeOutput(outputFile, "{\n")
 
+    increaseTabLevel()
+    # Wrap the call to WriteEvent with IsEnabled check
```
Comment (noahfalk) [inline src/coreclr/scripts/genRuntimeEventSources.py:159]: If I understood correctly all the events need to be hand-edited before they will handle Lttng correctly? If that is the case I think it would be much safer to generate a method body that throws NotImplementedException() and includes a comment describing what a future dev will need to change if they want it to have a correct implementation.

There is also a more minor concern that adding new auto-generated in the method body is making the size of System.Private.CoreLib.dll bigger storing code that will never be invoked.

**Review by noahfalk — COMMENTED** [review]
Glad to see this! A few questions and comments inline

File: src/libraries/System.Private.CoreLib/src/System/Threading/NativeRuntimeEventSource.PortableThreadPool.cs (hunk 5 above)
Comment (brianrob) [reply src/libraries/System.Private.CoreLib/src/System/Threading/NativeRuntimeEventSource.PortableThreadPool.cs:177]: You're right - this is calling `FireEtXPlat` which is beyond the decision of which logger to call - this will call LTTng on Linux directly.  You also bring up a good point that we may be double emitting.

Yes, I was proposing that we not worry about the difference in how the events are emitted when they are moved to managed code.  If we wanted to be more aggressive about maintaining perfect compatibility, we could also choose to build  auto-generated code that would allow us to call any of the events and have them emitted as a strongly typed LTTng event.  In the end, I think that the thing I want to encourage is that it would be nice if we didn't do special casing for these events here.  I suspect that we'll end up doing more of this type of thing, and it would be great if calling into `NativeRuntimeEventSource` to emit an event "just worked".

Comment (sywhang) [reply src/libraries/System.Private.CoreLib/src/System/Threading/NativeRuntimeEventSource.PortableThreadPool.cs:177]: Thanks for the suggestions @noahfalk and @brianrob. I went and checked and indeed I this causes it to double-emit events for EventListener. For LTTng it ends up being emitted as an "EventSourceEvent" event as well as the individual threadpool event. I'll fix this to use FireETW* events instead like @noahfalk suggested. 

> Yes, I was proposing that we not worry about the difference in how the events are emitted when they are moved to managed code.

I think this wouldn't really be a great experience for devs trying to parse an LTTng trace for a specific type of event by filtering via the name, so that was the main reason I added all these glue to make it seem indifferent from when we had the native threadpool. 

But I also agree with your sentiment towards having some kind of auto-generated code that'll allow devs to add events to ClrEtwAll.man and call it from managed side, and the whole thing "just works". I can try to tackle that later by modifying the scripts to autogenerate the QCall helpers and forward to corresponding LTTng sink.

File: src/libraries/System.Private.CoreLib/src/System/Threading/NativeRuntimeEventSource.PortableThreadPool.cs [hunk 7]
```diff
@@ -77,10 +59,6 @@ public enum ThreadAdjustmentReasonMap : uint
             ThreadTimedOut
         }
 
-        // Parameterized constructor to block initialization and ensure the EventSourceGenerator is creating the default constructor
-        // as you can't make a constructor partial.
-        private PortableThreadPoolEventSource(int _) { }
```
[the remark this answers is not in the document]
Comment (sywhang) [reply src/libraries/System.Private.CoreLib/src/System/Threading/NativeRuntimeEventSource.PortableThreadPool.cs:82]: This is now merged with NativeRuntimeEventSource, which already has this parameterized constructor.

File: src/libraries/System.Private.CoreLib/src/System/Threading/NativeRuntimeEventSource.PortableThreadPool.cs (hunk 3 above)
Comment (sywhang) [reply src/libraries/System.Private.CoreLib/src/System/Threading/NativeRuntimeEventSource.PortableThreadPool.cs:112]: Sorry, I don't think I quite got what you're asking here.. Do you mean what ensures that the DoCommand callback is received for the NativeRuntimeEventSource?

NativeRuntimeEventSource registers itself with the native runtime upon initialization, and it will receive callback from the runtime from the standard callback handler, just like how any other managed EventSources receive callbacks when being enabled.

Essentially the runtime "thinks" there are two providers -- the native provider whose callback is the EtwCallbackCommon function and the managed provider whose callback is the managed callback function in EventSource.

File: src/coreclr/scripts/genRuntimeEventSources.py (hunk 6 above)
Comment (sywhang) [reply src/coreclr/scripts/genRuntimeEventSources.py:159]: Ok, that makes sense. I'll modify this to throw instead :)

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/TraceLogging/XplatEventLogger.cs (hunk 2 above)
Comment (sywhang) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/TraceLogging/XplatEventLogger.cs:164]: This was preventing the provider to get enabled when only LTTng is specified via COMPlus_EnableEventLog=1. I moved this check to `OnEvenWritten` check instead so that they don't get forwarded as EventSourceEvent.

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/TraceLogging/XplatEventLogger.cs (hunk 4 above)
Comment (sywhang) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/TraceLogging/XplatEventLogger.cs]: The indirection is mainly because the QCall methods that call into the runtime need to be private, and the public wrappers are for the ThreadPool to be able to call them.

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs [hunk 8]
```diff
@@ -3804,11 +3805,8 @@ public class EventListener : IDisposable
         static EventListener()
         {
 #if FEATURE_PERFTRACING
-            // Ensure that NativeRuntimeEventSource is initialized so that EventListeners get an opportunity to subscribe to its events.
```
Comment (noahfalk) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs:3807]: Lets restore and update this comment. I assume the rationale is that we want EventListeners to be able to subscribe to events emitted by native code which may come before the first event emitted via managed code.

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/TraceLogging/XplatEventLogger.cs (hunk 4 above)
Comment (noahfalk) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/TraceLogging/XplatEventLogger.cs]: Are you sure? Here is an example of a QCall with the internal accesibility modifier: https://github.com/dotnet/runtime/blob/69e114c1abf91241a0eeecf1ecceab4711b8aa62/src/coreclr/System.Private.CoreLib/src/System/Runtime/InteropServices/NativeLibrary.CoreCLR.cs#L24

I wouldn't have expected QCall to care what accesibility modifier was assigned. If it does need to be non-public internal should still be sufficient to let the intra-assembly calls work without needing the wrapper method.

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/TraceLogging/XplatEventLogger.cs [hunk 9]
```diff
@@ -199,6 +201,119 @@ private static void LogOnEventWritten(EventWrittenEventArgs eventData)
 
             LogEventSource(eventData.EventId, eventData.EventName, eventData.EventSource.Name, payload);
         }
+#endif //FEATURE_EVENTSOURCE_XPLAT
+
+        [DllImport(RuntimeHelpers.QCall)]
```
Comment (noahfalk) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/TraceLogging/XplatEventLogger.cs:206]: Can we move all these QCalls definitions to the NativeRuntimeEventSource? They no longer have any clear relationship to XPlatEventLogger.

File: src/libraries/System.Private.CoreLib/src/System/Threading/NativeRuntimeEventSource.PortableThreadPool.cs [hunk 10]
```diff
@@ -1,28 +1,20 @@
...
-    //   use them to identify thread pool threads.
-    [EventSource(Name = "Microsoft-Windows-DotNETRuntime", Guid = "e13c0d23-ccbc-4e12-931b-d9cc2eee27e4")]
-    [EventSourceAutoGenerate]
-    internal sealed partial class PortableThreadPoolEventSource : EventSource
+    // This is part of the NativeRuntimeEventsource, which is the managed version of the Microsoft-Windows-DotNETRuntime provider.
+    // It contains the handwritten implementation of the ThreadPool events.
+    // The events here do not call into the typical WriteEvent* APIs unlike most EventSources.
+    // Instead, they call directly into QCalls provided by the runtime (refer to XplatEventLogger.cs) which call
+    // FireEtw* methods auto-generated from ClrEtwAll.man. This ensures that corresponding event sinks are being used
+    // for the native platform. Refer to src/coreclr/vm/nativeruntimesource.cpp.
```
Comment (noahfalk) [inline src/libraries/System.Private.CoreLib/src/System/Threading/NativeRuntimeEventSource.PortableThreadPool.cs:16]: I would clarify in the comment why WriteEvent() APIs are not being used here, that it causes double logging because of special case handling of this source in the EventListener code and gives us strongly typed LTTng events rather than the default LTTng handling that other EventSources get.

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/NativeRuntimeEventSource.cs [hunk 11]
```diff
@@ -5,19 +5,66 @@ namespace System.Diagnostics.Tracing
 {
     /// <summary>
     /// NativeRuntimeEventSource is an EventSource that represents the ETW/EventPipe events emitted by the native runtime.
-    /// Most of NativeRuntimeEventSource is auto-generated by scripts/genRuntimeEventSources.py based on the contents of the Microsoft-Windows-DotNETRuntime provider.
+    /// Most of NativeRuntimeEventSource is auto-generated by scripts/genRuntimeEventSources.py based on the contents of the Microsoft-Windows-DotNETRuntime provider and will throw a NotImplementedException without hand-written overloads of the Events
+    /// To have a runtime event be fired from the managed code, you need to add a managed definition for that event
+    /// and QCall into the native runtime that invoke the appropriate native sinks for the platform (i.e. ETW, EventPipe, LTTng).
+    /// To see some examples of this, refer to NativeRuntimeEventSource.PortableThreadPool.cs.
```
Comment (noahfalk) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/NativeRuntimeEventSource.cs:11]: Also we should add that they need to update the getRuntimeEventSources.py to make sure that script doesn't autogenerate a duplicate method that throws NotImplementedException.

**Review by noahfalk — APPROVED** [review]
I think its pretty close, I suggested a few more tweaks
